File: packages/android-sdk/src/main/kotlin/io/paaq/intelligence/PAAQ.kt
package io.paaq.intelligence

import android.app.Activity
import android.app.Application
import android.content.Context
import android.os.Build
import android.os.Bundle
import kotlinx.coroutines.*
import java.text.SimpleDateFormat
import java.util.*

/** Grace period after the app goes to background before a session is
 * considered ended — distinguishes a brief app-switch from the user
 * actually being done. */
private const val BACKGROUND_GRACE_MS = 30_000L

/**
 * PAAQ Intelligence SDK — Android
 *
 * Usage:
 *   class MyApp : Application() {
 *       override fun onCreate() {
 *           super.onCreate()
 *           PAAQ.initialize(this, sdkToken = "sdk_live_...", projectId = "proj_...")
 *       }
 *   }
 */
object PAAQ {

    private var api: PaaqApiClient? = null
    private var sessionId: String? = null
    private var sessionStartedAtMs: Long = 0
    private var sessionEnded = false
    private val queue = mutableListOf<PaaqEvent>()
    private val queueLock = Any()
    private var batchSize = 50
    private var config = PaaqConfig("", "")
    private val scope = CoroutineScope(Dispatchers.Default + SupervisorJob())
    private var flushJob: Job? = null
    private var heartbeatJob: Job? = null
    private var backgroundGraceJob: Job? = null
    private var activityCallbacks: Application.ActivityLifecycleCallbacks? = null
    private val isoFormat = SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ss.SSS'Z'", Locale.US).apply {
        timeZone = TimeZone.getTimeZone("UTC")
    }

    /**
     * Initialize the SDK. Call from Application.onCreate().
     */
    fun initialize(
        context: Context,
        sdkToken: String,
        projectId: String,
        environment: String = "production",
        debug: Boolean = false,
    ) {
        config = PaaqConfig(sdkToken, projectId, environment, debug)
        val deviceId = getOrCreateDeviceId(context)
        api = PaaqApiClient(config, deviceId)
        sessionEnded = false

        scope.launch {
            try {
                val result = api!!.sdkInit(collectDeviceMetadata(context))
                sessionId = result.sessionId
                sessionStartedAtMs = System.currentTimeMillis()
                result.batchSize?.let { batchSize = it }
                val flushInterval = result.flushIntervalSeconds ?: config.flushIntervalSeconds
                log("Initialized — session: $sessionId")
                startFlushLoop(flushInterval)
                startHeartbeatLoop(config.heartbeatIntervalSeconds)
                registerLifecycleCallbacks(context)
            } catch (e: Exception) {
                log("Init failed: ${e.message}")
            }
        }
    }

    private fun collectDeviceMetadata(context: Context): PaaqApiClient.DeviceMetadata {
        val appVersion = try {
            context.packageManager.getPackageInfo(context.packageName, 0).versionName
        } catch (_: Exception) { null }
        val metrics = context.resources.displayMetrics
        return PaaqApiClient.DeviceMetadata(
            osVersion = Build.VERSION.RELEASE ?: "",
            deviceModel = Build.MODEL ?: "",
            appVersion = appVersion,
            screenWidth = metrics.widthPixels,
            screenHeight = metrics.heightPixels,
        )
    }

    /**
     * Application.ActivityLifecycleCallbacks is a first-class, app-wide API
     * on Android (unlike iOS, which has no single equivalent) — safe to
     * install automatically for zero-touch screen tracking and real
     * background/foreground detection, no opt-in required.
     */
    private fun registerLifecycleCallbacks(context: Context) {
        val app = context.applicationContext as? Application ?: return
        activityCallbacks?.let { app.unregisterActivityLifecycleCallbacks(it) }

        var startedActivities = 0
        val callbacks = object : Application.ActivityLifecycleCallbacks {
            override fun onActivityCreated(activity: Activity, savedInstanceState: Bundle?) {}
            override fun onActivityResumed(activity: Activity) {
                screen(activity.javaClass.simpleName)
                installTouchTracking(activity)
            }
            override fun onActivityStarted(activity: Activity) {
                startedActivities++
                if (startedActivities == 1) {
                    backgroundGraceJob?.cancel()
                }
            }
            override fun onActivityStopped(activity: Activity) {
                startedActivities--
                if (startedActivities <= 0) {
                    backgroundGraceJob?.cancel()
                    backgroundGraceJob = scope.launch {
                        delay(BACKGROUND_GRACE_MS)
                        endSession("completed")
                    }
                }
            }
            override fun onActivityPaused(activity: Activity) {}
            override fun onActivitySaveInstanceState(activity: Activity, outState: Bundle) {}
            override fun onActivityDestroyed(activity: Activity) {}
        }
        activityCallbacks = callbacks
        app.registerActivityLifecycleCallbacks(callbacks)
    }

    private fun installTouchTracking(activity: Activity) {
        val window = activity.window ?: return
        val existing = window.callback
        if (existing is PaaqWindowCallbackWrapper) return // already wrapped
        val tracker = PaaqTouchTracker(
            onRageTap = { x, y, count -> track("\$rage_click", mapOf("x" to x, "y" to y, "tapCount" to count)) },
            onDeadTap = { x, y -> track("\$dead_click", mapOf("x" to x, "y" to y)) },
            lastSignalAtMs = { lastSignalAtMs },
        )
        window.callback = PaaqWindowCallbackWrapper(existing, tracker)
    }

    // ── Scroll depth — opt-in, no universal scroll hook exists on Android ──

    private var maxScrollPct = 0
    private val reportedScrollMilestones = mutableSetOf<Int>()

    /** Call from a scroll listener (RecyclerView.OnScrollListener, NestedScrollView.OnScrollChangeListener, etc.) with 0-100. */
    fun trackScrollDepth(pct: Int) {
        if (pct <= maxScrollPct) return
        maxScrollPct = pct
        for (milestone in intArrayOf(25, 50, 75, 100)) {
            if (pct >= milestone && reportedScrollMilestones.add(milestone)) {
                track("\$scroll_depth", mapOf("pct" to milestone))
            }
        }
    }

    /** Call when navigating to a new screen so scroll depth is measured per screen. */
    fun resetScrollTracking() {
        maxScrollPct = 0
        reportedScrollMilestones.clear()
    }

    // ── Form field friction — opt-in, wire into your existing focus/text listeners ──

    private val fieldStartedAtMs = mutableMapOf<String, Long>()
    private val fieldBackspaceCounts = mutableMapOf<String, Int>()

    fun trackFieldFocus(fieldName: String) {
        fieldStartedAtMs[fieldName] = System.currentTimeMillis()
        fieldBackspaceCounts[fieldName] = 0
    }

    fun trackFieldBackspace(fieldName: String) {
        fieldBackspaceCounts[fieldName] = (fieldBackspaceCounts[fieldName] ?: 0) + 1
    }

    fun trackFieldBlur(fieldName: String, formName: String? = null, hadError: Boolean = false, completed: Boolean = false) {
        val startedAt = fieldStartedAtMs.remove(fieldName)
        val backspaces = fieldBackspaceCounts.remove(fieldName) ?: 0
        track("\$form_field", mapOf(
            "fieldName" to fieldName,
            "formName" to (formName ?: ""),
            "timeSpentMs" to (startedAt?.let { System.currentTimeMillis() - it } ?: 0L),
            "backspaceCount" to backspaces,
            "hadError" to hadError,
            "completed" to completed,
        ))
    }

    fun trackFormAbandon(formName: String) {
        track("\$form_abandon", mapOf("formName" to formName))
    }

    private var lastSignalAtMs = 0L

    /** Track a custom event */
    fun track(eventName: String, properties: Map<String, Any> = emptyMap()) {
        lastSignalAtMs = System.currentTimeMillis()
        val event = PaaqEvent(eventName, sessionId, properties, now())
        synchronized(queueLock) { queue.add(event) }
        log("Queued $eventName")
        if (queueSize() >= batchSize) scope.launch { flush() }
    }

    /** Identify the current user */
    fun identify(userId: String, traits: Map<String, Any> = emptyMap()) {
        track("\$identify", traits + ("userId" to userId))
    }

    /** Record a screen view */
    fun screen(name: String) {
        track("\$screen", mapOf("name" to name))
    }

    /** Capture an exception */
    fun captureException(throwable: Throwable, context: Map<String, Any> = emptyMap()) {
        track("\$error", mapOf(
            "error_type" to throwable.javaClass.simpleName,
            "error_message" to (throwable.message ?: ""),
        ) + context)
    }

    /** Flush pending events to the backend */
    suspend fun flush() {
        val batch = synchronized(queueLock) {
            if (queue.isEmpty()) return
            val copy = queue.toList()
            queue.clear()
            copy
        }
        log("Flushing ${batch.size} events")
        api?.flush(batch)
    }

    /** Call from Application.onTerminate() or before process exit */
    fun dispose() {
        flushJob?.cancel()
        heartbeatJob?.cancel()
        backgroundGraceJob?.cancel()
        scope.launch {
            flush()
            endSession("completed")
        }
    }

    /** End the current session because the host app knows the user logged
     * out — the SDK cannot infer this on its own. */
    fun endSessionOnLogout() {
        scope.launch { endSession("logged_out") }
    }

    // Android's ActivityLifecycleCallbacks has no reliable signal for a hard
    // process kill (no equivalent to iOS's willTerminate) — that case is left
    // to session-sweep-cron server-side, which classifies a session that
    // goes silent as 'timed_out'/'abandoned'.
    private suspend fun endSession(outcome: String) {
        val id = sessionId ?: return
        if (sessionEnded) return
        sessionEnded = true
        val durationSeconds = if (sessionStartedAtMs > 0) {
            (System.currentTimeMillis() - sessionStartedAtMs) / 1000
        } else 0
        api?.endSession(id, durationSeconds, outcome)
    }

    private fun startFlushLoop(intervalSeconds: Long) {
        flushJob?.cancel()
        flushJob = scope.launch {
            while (isActive) {
                delay(intervalSeconds * 1000)
                flush()
            }
        }
    }

    private fun startHeartbeatLoop(intervalSeconds: Long) {
        heartbeatJob?.cancel()
        heartbeatJob = scope.launch {
            while (isActive) {
                delay(intervalSeconds * 1000)
                api?.heartbeat()
            }
        }
    }

    private fun queueSize(): Int = synchronized(queueLock) { queue.size }

    private fun now(): String = synchronized(isoFormat) { isoFormat.format(Date()) }

    private fun log(msg: String) {
        if (config.debug) android.util.Log.d("PAAQ", msg)
    }

    private fun getOrCreateDeviceId(context: Context): String {
        val prefs = context.getSharedPreferences("io.paaq.prefs", Context.MODE_PRIVATE)
        val key = "device_id"
        val existing = prefs.getString(key, null)
        if (existing != null) return existing
        val id = UUID.randomUUID().toString()
        prefs.edit().putString(key, id).apply()
        return id
    }
}

File: packages/ios-sdk/Sources/PaaqIntelligence/PAAQ.swift
import Foundation
#if canImport(UIKit)
import UIKit
#endif

/// PAAQ Intelligence SDK — iOS, macOS, tvOS, watchOS
public final class PAAQ {

    // MARK: - Public API

    /// Initialize the SDK. Call once, typically at app launch.
    @discardableResult
    public static func initialize(
        sdkToken: String,
        projectId: String,
        environment: String = "production",
        debug: Bool = false
    ) async -> Bool {
        return await shared.start(
            config: PaaqConfig(
                sdkToken: sdkToken,
                projectId: projectId,
                environment: environment,
                debug: debug
            )
        )
    }

    /// Track a custom event
    public static func track(_ eventName: String, properties: [String: Any] = [:]) {
        shared.enqueue(eventName: eventName, properties: properties)
    }

    /// Identify the current user
    public static func identify(_ userId: String, traits: [String: Any] = [:]) {
        var props = traits
        props["userId"] = userId
        shared.enqueue(eventName: "$identify", properties: props)
    }

    /// Record a screen view
    public static func screen(_ name: String) {
        shared.enqueue(eventName: "$screen", properties: ["name": name])
    }

    /// Capture an error
    public static func captureError(_ error: Error, context: [String: Any] = [:]) {
        var props: [String: Any] = [
            "error_type": String(describing: type(of: error)),
            "error_message": error.localizedDescription,
        ]
        props.merge(context) { $1 }
        shared.enqueue(eventName: "$error", properties: props)
    }

    /// Flush pending events to the backend
    public static func flush() async {
        await shared.flush()
    }

    /// Dispose and flush — call before app termination
    public static func dispose() async {
        await shared.teardown(outcome: "completed")
    }

    /// End the current session because the host app knows the user logged
    /// out — the SDK cannot infer this on its own.
    public static func endSessionOnLogout() async {
        await shared.teardown(outcome: "logged_out")
    }

    // MARK: - Private singleton

    private static let shared = PaaqInstance()
}

// MARK: - Internal actor

private actor PaaqInstance {
    private var api: PaaqApiClient?
    private var sessionId: String?
    private var sessionStartedAt: Date?
    private var sessionEnded = false
    private var explicitDispose = false
    private var queue: [PaaqEvent] = []
    private var batchSize = 50
    private var flushInterval: TimeInterval = 30
    private var flushTask: Task<Void, Never>?
    private var heartbeatTask: Task<Void, Never>?
    private var backgroundGraceTask: Task<Void, Never>?
    private var debug = false

    func start(config: PaaqConfig) async -> Bool {
        debug = config.debug
        let deviceId = getOrCreateDeviceId()
        let client = PaaqApiClient(config: config, deviceId: deviceId)
        api = client

        do {
            let result = try await client.sdkInit()
            sessionId = result.sessionId
            sessionStartedAt = Date()
            sessionEnded = false
            explicitDispose = false
            if let bs = result.serverBatchSize { batchSize = bs }
            if let fi = result.serverFlushInterval { flushInterval = fi }
            log("Initialized — session: \(sessionId ?? "nil")")
            startFlushLoop()
            startHeartbeatLoop()
            registerLifecycleObservers()
            return result.sessionId != nil
        } catch {
            log("Init failed: \(error)")
            return false
        }
    }

    func enqueue(eventName: String, properties: [String: Any]) {
        let props = properties.mapValues { AnyCodable($0) }
        queue.append(PaaqEvent(
            event_name: eventName,
            session_id: sessionId,
            properties: props,
            timestamp: ISO8601DateFormatter().string(from: Date())
        ))
        log("Queued \(eventName) (total: \(queue.count))")
        if queue.count >= batchSize { Task { await flush() } }
    }

    func flush() async {
        guard let api, !queue.isEmpty else { return }
        let batch = queue
        queue = []
        log("Flushing \(batch.count) events")
        await api.flush(events: batch)
    }

    func teardown(outcome: String) async {
        explicitDispose = true
        flushTask?.cancel()
        heartbeatTask?.cancel()
        backgroundGraceTask?.cancel()
        await flush()
        await endSession(outcome: outcome)
    }

    private func endSession(outcome: String) async {
        guard let api, let sessionId, !sessionEnded else { return }
        sessionEnded = true
        let duration = sessionStartedAt.map { Int(Date().timeIntervalSince($0)) } ?? 0
        await api.endSession(sessionId: sessionId, duration: duration, outcome: outcome)
    }

    private func startFlushLoop() {
        flushTask?.cancel()
        flushTask = Task {
            while !Task.isCancelled {
                try? await Task.sleep(nanoseconds: UInt64(flushInterval * 1_000_000_000))
                await flush()
            }
        }
    }

    private func startHeartbeatLoop() {
        heartbeatTask?.cancel()
        heartbeatTask = Task {
            while !Task.isCancelled {
                try? await Task.sleep(nanoseconds: 300_000_000_000) // 5 minutes
                guard let api else { continue }
                await api.heartbeat()
            }
        }
    }

    // Grace period after backgrounding before a session is considered ended —
    // distinguishes a brief app-switch from the user actually being done.
    private static let backgroundGracePeriod: UInt64 = 30_000_000_000 // 30s in nanoseconds

    private func registerLifecycleObservers() {
#if canImport(UIKit)
        NotificationCenter.default.addObserver(
            forName: UIApplication.willResignActiveNotification,
            object: nil,
            queue: nil
        ) { [weak self] _ in
            // Flush before iOS suspends the app
            Task { [weak self] in await self?.flush() }
        }

        NotificationCenter.default.addObserver(
            forName: UIApplication.didEnterBackgroundNotification,
            object: nil,
            queue: nil
        ) { [weak self] _ in
            Task { [weak self] in await self?.onBackgrounded() }
        }

        NotificationCenter.default.addObserver(
            forName: UIApplication.didBecomeActiveNotification,
            object: nil,
            queue: nil
        ) { [weak self] _ in
            Task { [weak self] in await self?.onForegrounded() }
        }

        NotificationCenter.default.addObserver(
            forName: UIApplication.willTerminateNotification,
            object: nil,
            queue: nil
        ) { [weak self] _ in
            // Real app termination — if dispose() wasn't called explicitly
            // first, this wasn't a deliberate teardown. Best-effort only:
            // iOS gives very little time to run code here.
            Task { [weak self] in await self?.onWillTerminate() }
        }
#endif
    }

    private func onBackgrounded() async {
        backgroundGraceTask?.cancel()
        backgroundGraceTask = Task { [weak self] in
            try? await Task.sleep(nanoseconds: PaaqInstance.backgroundGracePeriod)
            guard !Task.isCancelled else { return }
            await self?.endSession(outcome: "completed")
        }
    }

    private func onForegrounded() async {
        backgroundGraceTask?.cancel()
    }

    private func onWillTerminate() async {
        guard !explicitDispose else { return }
        await endSession(outcome: "force_closed")
    }

    private func log(_ message: String) {
        if debug { print("[PAAQ] \(message)") }
    }

    // MARK: - Device ID

    private func getOrCreateDeviceId() -> String {
        let key = "io.paaq.device_id"
        if let existing = UserDefaults.standard.string(forKey: key) { return existing }
        let id = UUID().uuidString
        UserDefaults.standard.set(id, forKey: key)
        return id
    }
}
